ic-agent/src/hash_tree.rs

```diff
@@ -0,0 +1,274 @@
+}
+
+impl fmt::Debug for Label {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fn printable(byte: u8) -> bool {
+            byte >= 32 && byte < 127
+        }
+        let bytes = self.as_bytes();
+        if bytes.iter().all(|b| printable(*b)) {
+            write!(f, "{}", std::str::from_utf8(bytes).unwrap())
+        } else {
+            write!(f, "0x")?;
+            bytes.iter().try_for_each(|b| write!(f, "{:02X}", b))
+        }
+    }
+}
+
+impl fmt::Display for Label {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl Into<String> for Label {
+    fn into(self) -> String {
+        format!("{}", self)
+    }
+}
+
+impl Into<serde_bytes::ByteBuf> for Label {
+    fn into(self) -> serde_bytes::ByteBuf {
+        serde_bytes::ByteBuf::from(self.to_vec())
+    }
+}
+
+impl<T> From<T> for Label
+where
+    T: std::convert::AsRef<[u8]>,
+{
+    fn from(bytes: T) -> Label {
+        let slice = bytes.as_ref();
+        let n = slice.len();
+        if n <= SMALL_LABEL_SIZE {
+            let mut buf = [0u8; SMALL_LABEL_SIZE + 1];
+            buf[0] = n as u8;
+            buf[1..=n].copy_from_slice(slice);
+            Self(LabelRepr::Value(buf))
+        } else {
+            Self(LabelRepr::Ref(slice.to_vec()))
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct Digest(pub [u8; 32]);
+
+impl Digest {
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+
+    pub fn to_vec(&self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+}
+
+impl fmt::Debug for Digest {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "0x")?;
+        self.0.iter().try_for_each(|b| write!(f, "{:02X}", b))
+    }
+}
+
+impl fmt::Display for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl From<[u8; 32]> for Digest {
+    fn from(bytes: [u8; 32]) -> Self {
+        Digest(bytes)
+    }
+}
+
+impl TryFrom<Vec<u8>> for Digest {
+    type Error = Vec<u8>;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        let a: Box<[u8; 32]> = bytes.into_boxed_slice().try_into()?;
+        Ok(Digest(*a))
+    }
+}
+
+impl AsRef<[u8]> for Digest {
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+pub struct Hasher(Sha256);
+
+impl Hasher {
+    pub fn for_domain(domain: &str) -> Self {
+        assert!(domain.len() < 256);
+        let mut hasher = Self(Sha256::new());
+        hasher.update(&[domain.len() as u8][..]);
+        hasher.update(domain.as_bytes());
+        hasher
+    }
+    pub fn update(&mut self, bytes: &[u8]) {
+        self.0.update(bytes);
+    }
+    pub fn finalize(self) -> Digest {
+        Digest(self.0.finish())
+    }
+}
+
+/// Public spec [hash-tree](https://hydra.dfinity.systems/latest/dfinity-ci-build/ic-ref.pr-218/public-spec/1/index.html#_encoding_of_certificates)
```

Comment:
Can we point the URL to the published (still internal) version of the spec that implements it? Something like https://docs.dfinity.systems/public/v/0.12.2/. Using a link to a rendered PR is bad form IMO (and the PR might be updated or rejected).

Reply:
There a small bits in that PR that are not in a  published version yet, waiting for https://github.com/dfinity-lab/dfinity/pull/6611/ to merge to do a release (we have fallen a bit into the habit of synchronizing the dfinity `master` merge with the `ic-ref` release). We should be able to update this later this week, I hope.

Reply:
The URL is not perfect yet.

But let's leave it here (it continues to have value for us, I think?), and update it later when we can.

It's okay if this PR is not perfect. It's okay if this PR is not perfect. It's okay if this PR is not perfect.